gitclone/gitclone.go

```diff
@@ -0,0 +1,226 @@
+}
+
+func exportEnvironmentWithEnvman(keyStr, valueStr string) error {
+	cmd := command.New("envman", "add", "--key", keyStr)
+	cmd.SetStdin(strings.NewReader(valueStr))
+	return cmd.Run()
+}
+
+func getMaxEnvLength() (int, error) {
+	configs, err := envman.GetConfigs()
+	if err != nil {
+		return 0, err
+	}
+
+	return configs.EnvBytesLimitInKB * 1024, nil
+}
+
+// Execute is the entry point of the git clone process
+func Execute(cfg Config) *StepError {
+	maxEnvLength, err := getMaxEnvLength()
+	if err != nil {
+		return NewStepError(
+			"get_max_commit_msg_length_failed",
+			fmt.Errorf("failed to set commit message length: %s", err),
+			"Getting allowed commit message length failed",
+		)
+	}
+
+	gitCmd, err := git.New(cfg.CloneIntoDir)
+	if err != nil {
+		return NewStepError(
+			"git_new",
+			fmt.Errorf("failed to create git project directory: %v", err),
+			"Creating new git project directory failed",
+		)
+	}
+	checkoutArg := getCheckoutArg(cfg.Commit, cfg.Tag, cfg.Branch)
+
+	originPresent, err := isOriginPresent(gitCmd, cfg.CloneIntoDir, cfg.RepositoryURL)
+	if err != nil {
+		return NewStepError(
+			"check_origin_present_failed",
+			fmt.Errorf("checking if origin is present failed: %v", err),
+			"Checking wether origin is present failed",
+		)
+	}
+
+	if originPresent && cfg.ResetRepository {
+		if err := resetRepo(gitCmd); err != nil {
+			return NewStepError(
+				"reset_repository_failed",
+				fmt.Errorf("reset repository failed: %v", err),
+				"Resetting repository failed",
+			)
+		}
+	}
+	if err := run(gitCmd.Init()); err != nil {
+		return NewStepError(
+			"init_git_failed",
+			fmt.Errorf("initializing repository failed: %v", err),
+			"Initializing git has failed",
+		)
+	}
+	if !originPresent {
+		if err := run(gitCmd.RemoteAdd("origin", cfg.RepositoryURL)); err != nil {
+			return NewStepError(
+				"add_remote_failed",
+				fmt.Errorf("adding remote repository failed (%s): %v", cfg.RepositoryURL, err),
+				"Adding remote repository failed",
+			)
+		}
+	}
+
+	isPR := cfg.PRRepositoryURL != "" || cfg.PRMergeBranch != "" || cfg.PRID != 0
+	if isPR {
+		if !cfg.ManualMerge || isPrivate(cfg.PRRepositoryURL) && isFork(cfg.RepositoryURL, cfg.PRRepositoryURL) {
+			if err := autoMerge(gitCmd, cfg.PRMergeBranch, cfg.BranchDest, cfg.BuildURL,
+				cfg.BuildAPIToken, cfg.CloneDepth, cfg.PRID); err != nil {
+				return NewStepError(
+					"auto_merge_failed",
+					fmt.Errorf("merging PR (automatic) failed: %v", err),
+					"Merging pull request failed",
+				)
+			}
+		} else {
+			if err := manualMerge(gitCmd, cfg.RepositoryURL, cfg.PRRepositoryURL, cfg.Branch,
+				cfg.Commit, cfg.BranchDest); err != nil {
+				return NewStepError(
+					"manual_merge_failed",
+					fmt.Errorf("merging PR (manual) failed: %v", err),
+					"Merging pull request failed",
+				)
+			}
+		}
+	} else if checkoutArg != "" {
+		if err := checkout(gitCmd, checkoutArg, cfg.Branch, cfg.CloneDepth, cfg.Tag != ""); err != nil {
+			return err
```

Comment:
should not this be `return NewStepError(...`

Reply:
checkout() returns a StepError.